[Script Telemetry] Add taintedness support for ScriptElement and CachedScript

whsieh opened the pull request:
#### 88d0219cb26e8e55dda7d3f3cfa8c94bce1d20c5
<pre>
[Script Telemetry] Add taintedness support for ScriptElement and CachedScript
<a href="https://bugs.webkit.org/show_bug.cgi?id=279601">https://bugs.webkit.org/show_bug.cgi?id=279601</a>

Reviewed by Abrar Rahman Protyasha, Charlie Wolfe, and Keith Miller.

Leverage JSC&apos;s taintedness tracking mechanism to propagate script telemetry state. See below for
more details.

* Source/JavaScriptCore/parser/SourceTaintedOrigin.cpp:
(JSC::sourceTaintedOriginFromStack):

Make this function return the closest tainted source URL along with the taintedness state, for use
in WebCore.

* Source/JavaScriptCore/parser/SourceTaintedOrigin.h:
* Source/JavaScriptCore/tools/JSDollarVM.cpp:
(JSC::JSC_DEFINE_HOST_FUNCTION):
* Source/WebCore/Headers.cmake:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/bindings/js/CachedScriptSourceProvider.h:
(WebCore::CachedScriptSourceProvider::CachedScriptSourceProvider):
* Source/WebCore/bindings/js/ScriptController.cpp:
(WebCore::ScriptController::executeJavaScriptURL):
* Source/WebCore/dom/ScriptElement.cpp:
(WebCore::ScriptElement::prepareScript):
(WebCore::ScriptElement::updateTaintedOriginFromSourceURL):

Teach `ScriptElement` to initialize `m_taintedOrigin` based on the source URL, immediately before
executing the script.

* Source/WebCore/dom/ScriptElement.h:
* Source/WebCore/dom/ScriptExecutionContext.cpp:
(WebCore::ScriptExecutionContext::requiresScriptExecutionTelemetry):
* Source/WebCore/dom/ScriptExecutionContext.h:
* Source/WebCore/loader/DocumentLoader.cpp:
(WebCore::DocumentLoader::handleContentFilterDidBlock):
* Source/WebCore/loader/cache/CachedResourceLoader.cpp:
(WebCore::scriptRequiresTelemetry):
(WebCore::createResource):
(WebCore::CachedResourceLoader::updateCachedResourceWithCurrentRequest):
(WebCore::CachedResourceLoader::requestResource):
(WebCore::CachedResourceLoader::revalidateResource):
(WebCore::CachedResourceLoader::loadResource):
* Source/WebCore/loader/cache/CachedScript.cpp:
(WebCore::CachedScript::CachedScript):
* Source/WebCore/loader/cache/CachedScript.h:

Add plumbing to set a tainted bit on `CachedScript` after it&apos;s done loading, so that we create the
corresponding `SourceOrigin` with the correct tainted bit set when evaluating it.

* Source/WebCore/page/ScriptTelemetryCategory.cpp: Added.
(WebCore::description):
* Source/WebCore/page/ScriptTelemetryCategory.h: Added.

Canonical link: <a href="https://commits.webkit.org/283713@main">https://commits.webkit.org/283713@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/396b39a06ee09b652da2e1299048f3f822ede2b9

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/67119 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/46498 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/19751 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/71154 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/18252 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/69237 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/54296 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/18045 "Built successfully") | [❌ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/53831 "Found 4 new test failures: fast/images/animated-jpegxl-loop-count.html fast/images/jpegxl-as-image.html fast/images/jpegxl-with-color-profile.html http/tests/images/repaint-garbled.html (failure)") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/12288 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/70186 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/42768 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/58079 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/34368 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/39440 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/15473 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/16606 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/60234 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/61405 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/15814 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/72857 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/66364 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/11078 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/15169 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/61302 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/11111 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/58138 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/61380 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/14863 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/9108 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/2714 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/88132 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/42303 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/15522 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/43380 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/44566 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/43121 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

kmiller68 reviewed (commented):
JSC changes look good to me.

aprotyas commented on Source/JavaScriptCore/tools/JSDollarVM.cpp:
```diff
@@ -2847,7 +2847,7 @@ JSC_DEFINE_HOST_FUNCTION(functionVMTaintedState, (JSGlobalObject* globalObject,
     VM& vm = globalObject->vm();
 
-    SourceTaintedOrigin sourceTaintedOrigin = sourceTaintedOriginFromStack(vm, callFrame);
+    auto [sourceTaintedOrigin, url] = sourceTaintedOriginFromStack(vm, callFrame);
```
We don't get an unused variable error for `url`?

n.b. In C++26 we should get the syntactic sugar to ignore `url` -- https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2020/p2169r0.pdf

whsieh replied:
Oh, apparently clang doesn't warn about unused variables as long as you use at least one of the results from the structured bindings assignment. I think I can make this robust if this ever changes, though, by just using `.first`:

```
auto sourceTaintedOrigin = sourceTaintedOriginFromStack(vm, callFrame).first;
```

kmiller68 replied:
It’s probably because structured binding “variables” aren’t actually variables and is why you can’t capture them in lambdas (in some cases I forget which though)

aprotyas commented on Source/WebCore/dom/ScriptExecutionContext.h:
```diff
@@ -57,6 +57,7 @@ class Exception;
 class JSPromise;
 class VM;
 enum class ScriptExecutionStatus;
+enum class SourceTaintedOrigin : uint8_t;
```
Is this forward decl needed?

whsieh replied:
Oh, it's not — good catch!

charliewolfe commented on Source/WebCore/dom/ScriptExecutionContext.h:
```diff
@@ -215,6 +216,8 @@ class ScriptExecutionContext : public SecurityContext, public TimerAlignment {
     WEBCORE_EXPORT void refAllowingPartiallyDestroyed();
     WEBCORE_EXPORT void derefAllowingPartiallyDestroyed();
 
+    WEBCORE_EXPORT bool requiresScriptExecutionTelemetry(ScriptTelemetryCategory);
```
Is this function supposed to be added in this patch? It’s not used.

whsieh replied:
Oh, it'll be used in the next patch, which will adopt this mechanism (along with the new enum type).

charliewolfe commented on Source/WebCore/loader/cache/CachedResourceLoader.cpp:
```diff
@@ -116,15 +116,27 @@ static inline ResourceErrorOr<CachedResourceHandle<T>> castCachedResourceTo(Reso
+    if (UNLIKELY(!document))
+        return ScriptRequiresTelemetry::No;
+
+    RefPtr page = document->protectedPage();
```
nit - unneeded reference churn
```suggestion
    RefPtr page = document->page();
```

whsieh replied:
Fixed!

charliewolfe commented on Source/WebCore/dom/ScriptExecutionContext.cpp:
```diff
@@ -957,6 +959,21 @@ RefCountedSerialFunctionDispatcher& ScriptExecutionContext::nativePromiseDispatc
 
+bool ScriptExecutionContext::requiresScriptExecutionTelemetry(ScriptTelemetryCategory category)
+{
+    Ref vm = protectedVM();
```
nit - unneeded reference churn
```suggestion
    Ref vm = this->vm();
```

charliewolfe commented on Source/WebCore/dom/ScriptExecutionContext.cpp:
```diff
@@ -957,6 +959,21 @@ RefCountedSerialFunctionDispatcher& ScriptExecutionContext::nativePromiseDispatc
+    UNUSED_PARAM(taintedURL);
+    UNUSED_PARAM(category);
+
+    return taintedness == JSC::SourceTaintedOrigin::KnownTainted;
```
This function isn’t used right now, but in the future won’t we also want this to return true when `taintedness == JSC::SourceTaintedOrigin::IndirectlyTainted`?

whsieh replied:
Ah, I left it `KnownTainted` for now to be a bit more conservative, but I can go ahead and include this now.

whsieh committed:
[Script Telemetry] Add taintedness support for ScriptElement and CachedScript
https://bugs.webkit.org/show_bug.cgi?id=279601

Reviewed by Abrar Rahman Protyasha, Charlie Wolfe, and Keith Miller.

Leverage JSC's taintedness tracking mechanism to propagate script telemetry state. See below for
more details.

* Source/JavaScriptCore/parser/SourceTaintedOrigin.cpp:
(JSC::sourceTaintedOriginFromStack):

Make this function return the closest tainted source URL along with the taintedness state, for use
in WebCore.

* Source/JavaScriptCore/parser/SourceTaintedOrigin.h:
* Source/JavaScriptCore/tools/JSDollarVM.cpp:
(JSC::JSC_DEFINE_HOST_FUNCTION):
* Source/WebCore/Headers.cmake:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/bindings/js/CachedScriptSourceProvider.h:
(WebCore::CachedScriptSourceProvider::CachedScriptSourceProvider):
* Source/WebCore/bindings/js/ScriptController.cpp:
(WebCore::ScriptController::executeJavaScriptURL):
* Source/WebCore/dom/ScriptElement.cpp:
(WebCore::ScriptElement::prepareScript):
(WebCore::ScriptElement::updateTaintedOriginFromSourceURL):

Teach `ScriptElement` to initialize `m_taintedOrigin` based on the source URL, immediately before
executing the script.

* Source/WebCore/dom/ScriptElement.h:
* Source/WebCore/dom/ScriptExecutionContext.cpp:
(WebCore::ScriptExecutionContext::requiresScriptExecutionTelemetry):
* Source/WebCore/dom/ScriptExecutionContext.h:
* Source/WebCore/loader/DocumentLoader.cpp:
(WebCore::DocumentLoader::handleContentFilterDidBlock):
* Source/WebCore/loader/cache/CachedResourceLoader.cpp:
(WebCore::scriptRequiresTelemetry):
(WebCore::createResource):
(WebCore::CachedResourceLoader::updateCachedResourceWithCurrentRequest):
(WebCore::CachedResourceLoader::requestResource):
(WebCore::CachedResourceLoader::revalidateResource):
(WebCore::CachedResourceLoader::loadResource):
* Source/WebCore/loader/cache/CachedScript.cpp:
(WebCore::CachedScript::CachedScript):
* Source/WebCore/loader/cache/CachedScript.h:

Add plumbing to set a tainted bit on `CachedScript` after it's done loading, so that we create the
corresponding `SourceOrigin` with the correct tainted bit set when evaluating it.

* Source/WebCore/page/ScriptTelemetryCategory.cpp: Added.
(WebCore::description):
* Source/WebCore/page/ScriptTelemetryCategory.h: Added.

Canonical link: https://commits.webkit.org/283713@main
